docs/reference/transforms/after_deps.rst

```diff
@@ -0,0 +1,148 @@
+
+In this example, a task will be created for each build *and* each toolchain.
+
+Limiting Dependencies by Attribute
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+It may not be desirable to create a new task for *all* tasks in the given kind.
+It's possible to limit the tasks given some attribute:
+
+.. code-block:: yaml
+
+   kind-dependencies:
+     - build
+
+   tasks:
+     signing:
+       after-deps:
+         with-attributes:
+           platform: linux
+
+In the above example, follow-up tasks will only be created for builds whose
+``platform`` attribute equals "linux". Multiple attributes can be specified,
+these are resolved using the :func:`~taskgraph.util.attributes.attrmatch`
+utility function.
+
+Grouping Dependencies
+~~~~~~~~~~~~~~~~~~~~~
+
+Sometimes, it may be desirable to run a task after a group of related tasks
+rather than an individual task. One example of this, is a task to notify when a
+release is finished for each platform we're shipping. We want it to depend on
+every task from that particular release phase.
+
+To accomplish this, we specify the ``group-by`` key:
+
+.. code-block:: yaml
+
+   kind-dependencies:
+     - build
+     - signing
+     - publish
+
+   tasks:
+     notify:
+       after-deps:
+         group-by:
+           attribute: platform
+
+In this example, tasks across the ``build``, ``signing`` and ``publish`` kinds will
+be scanned for an attribute called "platform" and sorted into corresponding groups.
+Assuming we're shipping on Windows, Mac and Linux, it might create the following
+groups:
+
+.. code-block::
+
+   - build-windows, signing-windows, publish-windows
+   - build-mac, signing-mac, publish-mac
+   - build-linux, signing-linux, publish-linux
+
+Then the ``notify`` task will be duplicated into three, one for each group. The
+notify tasks will depend on each task in its associated group.
+
+Custom Grouping
+~~~~~~~~~~~~~~~
+
+Only the default ``single`` and the ``attribute`` group-by functions are
+built-in. But if more complex grouping is needed, custom functions can be
+implemented as well:
+
+.. code-block:: python
+
+   from typing import List
+
+   from taskgraph.task import Task
+   from taskgraph.transforms.after_deps import group_by
+   from taskgraph.transforms.base import TransformConfig
+
+   @group_by("custom-name")
+   def group_by(config: TransformConfig, tasks: List[Task]) -> List[List[Task]]:
+      pass
+
+It's also possible to specify a schema for your custom group-by function, which
+allows tasks to pass down additional context (such as with the built-in
+``attribute`` function):
+
+.. code-block:: python
+
+   from typing import List
+
+   from taskgraph.task import Task
+   from taskgraph.transforms.after_deps import group_by
+   from taskgraph.transforms.base import TransformConfig
+   from taskgraph.util.schema import Schema
+
+   @group_by("custom-name", schema=Schema(str))
+   def group_by(config: TransformConfig, tasks: List[Task], ctx: str) -> List[List[Task]]:
```

Comment:
It would be nice to include an example of defining this extra context (I'm assuming that happens in the kind using `after_deps`?). I'm not sure how it would work based on this.